user_tools/src/spark_rapids_tools/storagelib/tools/fs_utils.py

```diff
@@ -0,0 +1,186 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility functions for file system operations"""
+
+import dataclasses
+import tarfile
+from functools import cached_property
+from typing import Optional, Union, List
+
+from pydantic import model_validator, FilePath, AnyHttpUrl, StringConstraints, ValidationError
+from pydantic.dataclasses import dataclass
+from pydantic_core import PydanticCustomError
+from typing_extensions import Annotated
+
+from spark_rapids_tools import CspPath
+from spark_rapids_tools.enums import HashAlgorithm
+from spark_rapids_tools.storagelib import LocalPath
+
+CspPathString = Annotated[str, StringConstraints(pattern=r'^\w+://.*')]
+"""
+A type alias for a string that represents a path. The path must start with a protocol scheme.
+"""
+
+
+def strip_extension(file_name: str, count: int = 1) -> str:
+    """
+    Utility method to strip the extension from a file name. By default, it only removes the last extension.
+    The caller can override the count of extensions. For examples:
+    :param file_name: The file name.
+    :param count: The number of extensions to remove. i.e., this for files with multi-extensions.
+    :return: The file name without the extension.
+
+    Examples:
+    ```py
+    strip_extension('foo.tgz')                  # returns 'foo'
+    strip_extension('foo.tar.gz', count = 2)    # returns 'foo'
+    strip_extension('foo.tar.gz')               # returns 'foo.tar'
+    ```
+    """
+    return file_name.rsplit('.', count)[0]
+
+
+def raise_invalid_file(file_path: CspPath, msg: str, error_type: str = 'invalid_file') -> None:
+    """
+    Utility method to raise a custom pydantic error for invalid files. See the custom pydantic
+    errors for more details: https://docs.pydantic.dev/latest/errors/errors/#custom-errors
+    :param file_path: The object instantiated with the file path.
+    :param msg:
+    :param error_type: the error type top be displayed in the error message.
+    :return:
+    """
+    raise PydanticCustomError(error_type, f'File {str(file_path)} {msg}')
+
+
+@dataclass
+class FileHashAlgorithm:
+    """
+    Represents a file hash algorithm and its value. Used for verification against an
+    existing file.
+    ```py
+    try:
+        file_algo = FileHashAlgorithm(algorithm=HashAlgorithm.SHA256, value='...')
+        file_algo.verify_file(CspPath('file://path/to/file'))
+    except ValidationError as e:
+        print(e)
+    ```
+    """
+    algorithm: HashAlgorithm
+    value: str
+
+    def verify_file(self, file_path: CspPath) -> bool:
+        cb = self.algorithm.get_hash_func()
+        with file_path.open_input_stream() as stream:
+            res = cb(stream.readall())
```

Comment:
I assume readall is going to pull it all into memory?  really large files might have an issue with that.

Reply:
that's correct. It is tradeoff with memory.
I was fine with that tradeoff because the download task only runs as part of initializing the environment. At this stage the memory should not be stressed by the running the tools.
The fat-wheel script also runs the download-tasks but this is mainly a standalone script and should not be causing any runtime issues for the users.